File: NeetCodeIO/206ReverseLinkedList/ReverseLinkedList.py
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # 3 pointers -> Time:O(n), Space:O(1)
        if not head:
            return head
        prev, curr, nxt = None, head, head.next
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        return prev

File: NeetCodeIO/206ReverseLinkedList/ReverseLinkedList.cpp
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        // 3 pointers -> Time:O(n), Space:O(1)
        if (!head) {
            return head;
        }
        ListNode *prev = NULL, *curr = head, *next = head->next;
        while (curr) {
            next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }
};
